controlplane/internal/controller/agentcontrolplane_controller.go

```diff
@@ -199,6 +206,43 @@ func (r *AgentControlPlaneReconciler) Reconcile(ctx context.Context, req ctrl.Re
 	return ctrl.Result{}, rerr
 }
 
+// Ensures dependencies are deleted before allowing the AgentControlPlane to be deleted
+// Deletes the ClusterDeployment (which deletes the AgentClusterInstall)
+// Machines, InfraMachines, and AgentBootstrapConfigs get auto-deleted when the ACP has a deletion timestamp - this deprovisions the BMH automatically
+// TODO: should we handle watching until all machines & agentbootstrapconfigs are deleted too?
+func (r *AgentControlPlaneReconciler) handleDeletion(ctx context.Context, acp *controlplanev1beta1.AgentControlPlane) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+	if controllerutil.ContainsFinalizer(acp, acpFinalizer) {
+		log.Info("ACP contains finalizer, handling deletion")
+		// Delete cluster deployment
+		if acp.Status.ClusterDeploymentRef != nil {
+			cd := &hivev1.ClusterDeployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      acp.Status.ClusterDeploymentRef.Name,
+					Namespace: acp.Status.ClusterDeploymentRef.Namespace,
+				},
+			}
+			if err := r.Client.Delete(ctx, cd); err != nil && !apierrors.IsNotFound(err) {
+				log.Error(err, "failed deleting cluster deployment for ACP")
+				return ctrl.Result{}, err
+			}
```

Comment:
We can extract this into `deleteClusterDeployment(acp.StatusClusterDeploymentRef)`

Reply:
Oh yes sorry I forgot about this, will modify!

Reply:
Changed it!